File: ocaml/bin/retirement_calculator/retirement_calculator.ml
open CalendarLib

let current_year () =
  let now = Calendar.now () in
  Calendar.year now

let rec prompt_for_input prompt =
  print_string prompt;
  let input = read_line () in
  match int_of_string_opt input with
  | Some int_input when int_input < 1 ->
      print_endline "Error: Please provide a number greater than or equal to 1.";
      prompt_for_input prompt
  | Some int_input -> int_input
  | None ->
      if input = "" then
        print_endline "Error: Please provide a non-blank input."
      else print_endline "Error: Please provide a valid number.";
      prompt_for_input prompt

let () =
  let age = prompt_for_input "What is your current age?\n" in
  let retirement_age =
    prompt_for_input "What age would you like to retire?\n"
  in

  if retirement_age > age then
    let year = current_year () in
    let years_left = retirement_age - age in
    let retirement_year = year + years_left in
    Printf.printf
      "You have %i years left until you retire.\n\
       It's %i, so you can retire in %i.\n"
      years_left year retirement_year
  else Printf.printf "You can already retire!\n"

File: typescript/src/retirementCalculator.ts
import * as readline from "readline";

export function isValidAnswer(answer: string) {
  if (answer.trim() === "") {
    return {
      isValid: false,
      error: "Input cannot be empty. Please try again.",
    };
  } else if (isNaN(parseInt(answer))) {
    return {
      isValid: false,
      error: "Input must be a number. Please try again.",
    };
  }
  return { isValid: true, error: "" }; // No error if valid
}

// Step 1: Input
function getInput(question: string): Promise<number> {
  return new Promise((resolve) => {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    const askQuestion = () => {
      rl.question(question, (answer) => {
        const validation = isValidAnswer(answer);
        if (validation.isValid) {
          rl.close();
          resolve(parseInt(answer));
        } else {
          console.log(validation.error);
          askQuestion(); // Recursively ask the question again if validation fails
        }
      });
    };

    askQuestion(); // Initial question
  });
}

// Step 2: Concatenation
export function createResponse(
  currentAge: number,
  retirementAge: number,
): string {
  const yearsLeft = retirementAge - currentAge;
  const currentYear = new Date().getFullYear();
  const retireYear = currentYear + yearsLeft;
  return `You have ${yearsLeft} years left until you can retire.\nIt's ${currentYear}, so you can retire in ${retireYear}.`;
}

// Step 3: Output
function printResponse(message: string): void {
  console.log(message);
}

// Main function to run the program
async function main() {
  const currentAge = await getInput("What is your current age? ");
  const retirementAge = await getInput(
    "At what age would you like to retire? ",
  );
  const response = createResponse(currentAge, retirementAge); // Create greeting message
  printResponse(response); // Print greeting message
}

main();
